src/kafkaaggregator/testtopic/agents.py

```diff
@@ -0,0 +1,157 @@
+"""Faust agent for the test topic aggregation example.
+
+The Faust agent persists the messages for the test topic in tumbling windows.
+See https://faust.readthedocs.io/en/latest/userguide/tables.html#windowing
+
+The window_size parameter controls the number of messages to aggregate,
+if the test topic is produced at a constant frequency, the number of messages
+aggregated in each window is n = window_size * frequency. The window range is
+relative to the time field in the test topic.
+
+The window_expires parameter controls when the callback function to
+process the expired window(s) is called. When a window is processed, a new
+message for the aggregated test topic is produced with the aggregated results.
+"""
+
+__all__ = ["aggregate", "process_window", "process_stream"]
+
+
+import logging
+from statistics import mean
+from typing import AsyncGenerator, List, Tuple
+
+from faust import web
+from faust.types import StreamT
+
+from kafkaaggregator.app import app
+from kafkaaggregator.config import Configuration
+from kafkaaggregator.testtopic.models import AggTestTopic, TestTopic
+
+logger = logging.getLogger("kafkaaggregator")
+
+config = Configuration()
+
+test_topic = app.topic("test-topic", value_type=TestTopic, internal=True)
+
+aggregated_test_topic = app.topic(
+    "agg-test-topic", value_type=AggTestTopic, internal=True
+)
+
+
+def aggregate(timestamp: float, messages: List[TestTopic]) -> AggTestTopic:
+    """Return an aggregated message from a list of messages for the test
+    topic and a timestamp.
+
+    Parameters
+    ----------
+
+    timestamp: `float`
+        A timestamp for the aggregated message, for example, the mid point of
+        the window that contains the messages.
+    messages: `List[TestTopic]`
```

Comment:
Note you technically can't use `List[TestTopic]` within single backticks because that's not a Python type (it's more a Python expression). It's safer to leave it as `list` in the docstring, or see https://developer.lsst.io/python/numpydoc.html#sequence-types

Reply:
Yep, using static typing is something new to me will make sure I get the docstrings right :)